dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/sync/CompleteDataSetRegistrationSynchronization.java

```diff
@@ -27,83 +27,121 @@
-                "Complete data set registration synchronization failed. Remote " +
-                    "server is unavailable." );
+                "Complete data set registration synchronization failed. Remote server is unavailable." );
         }
-        final Clock clock = new Clock( log ).startClock()
-            .logTime( "Starting Complete data set registration synchronization job." );
-
-        // ---------------------------------------------------------------------
-        // Set time for last success to start of process to make data saved
-        // subsequently part of next synch process without being ignored
-        // ---------------------------------------------------------------------
-        ImportSummary importSummary;
-        try
-        {
-            importSummary = synchronizationManager.executeCompleteDataSetRegistrationPush();
-            if ( SyncUtils
-                .checkSummaryStatus( importSummary, SyncEndpoint.COMPLETE_DATA_SET_REGISTRATIONS ) )
-            {
-                String resultMsg = "Complete data set registration synchronization job is done. It took ";
-                clock.logTime( "SUCCESS! " + resultMsg );
-                SyncUtils.setLastSyncSuccess( systemSettingManager,
-                    SettingKey.LAST_SUCCESSFUL_COMPLETE_DATA_SET_REGISTRATION_SYNC,
-                    new Date( clock.getStartTime() ) );
-                return SynchronizationResult
-                    .newSuccessResultWithMessage( resultMsg + clock.getTime() + " ms." );
-            }
 
+        initializeSyncVariables();
+
+        if ( objectsToSynchronize == 0 )
+        {
+            SyncUtils.setLastSyncSuccess( systemSettingManager,
+                SettingKey.LAST_SUCCESSFUL_COMPLETE_DATA_SET_REGISTRATION_SYNC, new Date( clock.getStartTime() ) );
+            log.info( "Skipping completeness synchronization, no new or updated data" );
+            return SynchronizationResult.newSuccessResultWithMessage( "Skipping completeness synchronization, no new or updated data" );
         }
-        catch ( Exception ex )
+
+        if ( sendSyncRequest() )
         {
-            log.error(
-                "Exception happened while trying complete data set registration push " + ex.getMessage(),
-                ex );
+            String resultMsg = "Complete data set registration synchronization is done. It took ";
+            clock.logTime( "SUCCESS! " + resultMsg );
+            SyncUtils.setLastSyncSuccess( systemSettingManager,
+                SettingKey.LAST_SUCCESSFUL_COMPLETE_DATA_SET_REGISTRATION_SYNC,
+                new Date( clock.getStartTime() ) );
+
+            return SynchronizationResult.newSuccessResultWithMessage( resultMsg + clock.getTime() + " ms." );
         }
+
         return SynchronizationResult
             .newFailureResultWithMessage( "Complete data set registration synchronization failed." );
     }
+
+    private void initializeSyncVariables()
+    {
```

Comment:
This private initialization may be hard to unit test, especially because static calls are used (`SyncUtils.getLastSyncSuccess(..)`)